feat(amazonq): /doc: add support for infrastructure diagrams

vikshe opened the pull request:
Amazon Q /doc: Add support for infrastructure diagrams


![image](https://github.com/user-attachments/assets/2ab56138-d65d-4036-977c-d5caafc68559)



- Treat all work as PUBLIC. Private `feature/x` branches will not be squash-merged at release time.
- Your code changes must meet the guidelines in [CONTRIBUTING.md](https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#guidelines).
- License: I confirm that my contribution is made under the terms of the Apache 2.0 license.

Commit:
feat(amazonq): /doc: add support for infrastructure diagrams

avi-alpert reviewed (changes requested):
adm-zip is no longer being used, see https://github.com/aws/aws-toolkit-vscode/commit/21c7f12132dc4cc72724e1946dc0ce1ea627edc0
thanks, updated this PR

avi-alpert commented on packages/core/src/amazonqDoc/session/sessionState.ts:
```diff
@@ -363,7 +364,9 @@ export class PrepareCodeGenState implements SessionState {
                 action.telemetry,
-                span
+                span,
+                new AdmZip(),
```
admzip is no longer being used

vikshe replied:
thanks. Updated

avi-alpert commented on packages/core/src/amazonqFeatureDev/util/files.ts:
```diff
@@ -26,6 +30,15 @@ export async function checkForDevFile(root: string) {
+function isInfraDiagramFile(relativePath: string) {
+    return (
+        relativePath.toLowerCase() === 'infra.dot' ||
+        relativePath.toLowerCase().endsWith('/infra.dot') ||
```
relativeFilePath on windows might use backwards slashes

vikshe replied:
updated

avi-alpert commented on packages/core/src/amazonqFeatureDev/util/files.ts:
```diff
@@ -26,6 +30,15 @@ export async function checkForDevFile(root: string) {
 
+function isInfraDiagramFile(relativePath: string) {
+    return (
+        relativePath.toLowerCase().endsWith('docs/infra.dot') ||
```
`path.join` could be used here instead

avi-alpert commented on packages/core/src/amazonqFeatureDev/util/files.ts:
```diff
@@ -34,13 +47,38 @@ export async function prepareRepoData(
         // We only respect gitignore file rules if useAutoBuildFeature is on, this is to avoid dropping necessary files for building the code (e.g. png files imported in js code)
-        const files = await collectFiles(repoRootPaths, workspaceFolders, true, maxRepoSizeBytes, !useAutoBuildFeature)
+        if (!useAutoBuildFeature) {
+            if (isIncludeInfraDiagram) {
```
is there a way we could avoid the additional special case logic here? could we either a) remove svg from the exclude patterns, or b) pass in true for defaultExcludePatterns (like /dev does with useAutoBuildFeature). with the addition of isIncludeInfraDiagram its hard to follow all the different scenarios

avi-alpert commented on packages/core/src/shared/utilities/workspaceUtils.ts:
```diff
@@ -313,30 +328,26 @@ async function filterOutGitignoredFiles(
+}
+export type CollectFilesFileFilter = (relativePath: string) => boolean // returns true if file should be filtered out
+
+export async function collectFiles2(
```
The name of this function doesn't add value, why is there a second collect files function? take a look at naming section in docs/CODE_GUIDELINES.md

avi-alpert replied:
It looks like the intention here is to have collectFiles use a labeled options object instead of individual parameters, if thats the case, perhaps collectFiles itself should be changed instead of adding a new function

zuoyaofu committed:
fix(amazonq): Auto-trigger review flow on scan button click

laileni-aws committed:
fix(amazonq): /test may truncate user prompt #6557

## Problem
- If user gives user input prompt more than 4096 characters, this fails
at RTS as model supports only 4096 input characters.
`Value at 'userInput' failed to satisfy constraint: Member must have
length less than or equal to 4096`
- Ideally we should not get more than 4096 char input from user input
but still features like `/test`, `/dev` were able to get the message
length more than 4096.

## Solution
- To resolve this issue, we are adding a truncation logic to continue
the test generation workflow instead of failing at RTS.
- No new tests were added as the functionality is not impacted.

zuoyaofu committed:
Merge branch 'master' into review-ui-bug

zuoyaofu committed:
Update packages/amazonq/.changes/next-release/Bug Fix-e9cfb8e0-540c-45ec-b3b7-313ac07c87a7.json

Co-authored-by: Tai Lai <ctlai95@gmail.com>

nkomonen-amazon committed:
fix(build): webviews sometimes did not load during development (#6556)

## Problem:

In our development build we spawn a webpack server that serves the Vue
files used by our webviews. It is served on port 8080, but sometimes an
existing process is already using it, and this causes the webpack server
to use the next available port. But because we assume it will use 8080
it silently breaks, and when developing the Q webview would not load as
expected.

## Solution:

Explicitly require 8080 in the webpack server cli command, also create a
custom script to catch the error early and explain how to resolve it.

Now in the build tasks if 8080 is in use the dev can see the failed task
with the solution explicitly stated.

---

- Treat all work as PUBLIC. Private `feature/x` branches will not be
squash-merged at release time.
- Your code changes must meet the guidelines in
[CONTRIBUTING.md](https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#guidelines).
- License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

Signed-off-by: nkomonen-amazon <nkomonen@amazon.com>

dhasani23 committed:
fix(amazonq): let chat update after entering jdk path (#6562)

## Problem

After entering JDK path, chat is sometimes frozen until the local build
completes.


## Solution

Pause after user enters JDK path, to give chat time to update before
starting the local build.

---

- Treat all work as PUBLIC. Private `feature/x` branches will not be
squash-merged at release time.
- Your code changes must meet the guidelines in
[CONTRIBUTING.md](https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#guidelines).
- License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

Co-authored-by: David Hasani <davhasan@amazon.com>

Will-ShaoHua committed:
config(customization): customization override (#6545)

## Problem

https://github.com/aws/aws-toolkit-vscode/blob/master/packages/core/src/codewhisperer/util/customizationUtil.ts#L116-L130

The flaw of previous implementation won't override customization if
users have selected one.

The team is asking 
1. Ability to override regardless users are with a customization or not
2. Should respect users' selection afterward when users switch back to
the customization they prefer
3. When ab service provides a new override customization arn, plugins
should be able to do override again

```
// as long as there is a customization selected, override will not happen no matter what
 if (selectedCustomization && selectedCustomization.name !== '') {
        return selectedCustomization
    } else {
        const customizationFeature = FeatureConfigProvider.getFeature(Features.customizationArnOverride)
        const arnOverride = customizationFeature?.value.stringValue
        const customizationOverrideName = customizationFeature?.variation
        if (arnOverride === undefined) {
            return baseCustomization
        } else {
            return {
                arn: arnOverride,
                name: customizationOverrideName,
                description: baseCustomization.description,
            }
        }

```

## Solution
1. Persist a value of "previous override arn" and do override when the
provided customization arn is different than "previous override arn"



---

- Treat all work as PUBLIC. Private `feature/x` branches will not be
squash-merged at release time.
- Your code changes must meet the guidelines in
[CONTRIBUTING.md](https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#guidelines).
- License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

chungjac committed:
telemetry(amazonq): add feedback functionality with thumbs up/down (#6560)

## Problem
- /test and /review are missing thumbs up/down telemetry
- /test is missing feedback functionality

<img width="545" alt="Screenshot 2025-02-11 at 1 58 42 PM"
src="https://github.com/user-attachments/assets/11a46136-46a6-4d35-a7e1-e92f45b1034a"
/>
<img width="558" alt="Screenshot 2025-02-11 at 1 58 48 PM"
src="https://github.com/user-attachments/assets/80288c37-a98c-4e8f-9e32-810238ee2695"
/>


## Solution
- send telemetry event for thumps up/down for /test and /review
- send user feedback for /test which routes to SIM folder
- No tests were added as this is UX change

---

- Treat all work as PUBLIC. Private `feature/x` branches will not be
squash-merged at release time.
- Your code changes must meet the guidelines in
[CONTRIBUTING.md](https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#guidelines).
- License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

zuoyaofu committed:
Merge pull request #6550 from zuoyaofu/review-ui-bug

fix(amazonq): Auto-trigger review flow on scan button click

jonife committed:
perf(appbuilder): reduce number of ListRemoteTestEvent calls by remote invoke #6559

Problem:
Remote Invoke call ListRemoteTestEvent every time the webview is instantiated. 

Solution:
Introduce a condition for which this api calls can be
made thereby reducing the spike in number failures we see in our
dashboard.

dhasani23 committed:
fix(amazonq): make sure target JDK version is defined (#6569)

## Problem

Due to Java 21 support coming out, the target JDK version was no longer
hardcoded as JDK17. So for SQL conversions, the target JDK was
`undefined` and the `startJob` API was failing.


## Solution

Hardcode JDK17 as a dummy target JDK so that we don't encounter an
invalid request exception.

---

- Treat all work as PUBLIC. Private `feature/x` branches will not be
squash-merged at release time.
- Your code changes must meet the guidelines in
[CONTRIBUTING.md](https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#guidelines).
- License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

Co-authored-by: David Hasani <davhasan@amazon.com>

Hweinstock committed:
ci(jscpd): add hint message for rebasing on false negatives #6564

## Problem
The duplicate code detection can fail when the current branch is not
rebased with the most recent version of master. This is a fairly common
false-negative. To avoid this, the PR must be kept up to date with its
target branch. However, the tool doesn't provide any hints/indication
this is the case.

## Solution
- add a hint message on failure that suggests rebasing.

evanliu048 committed:
refactor(amazonq): revert pre-fetch next recommendations for inline completion (#6567)

This reverts commit 50c91210736c1226e8bcb4c1af10fe29fe202ee7.

## Problem
In the previous change, we applied prefect to every responses. However,
this introduced issues with inline completion, causing it to behave
inconsistently and produce unexpected completions in the next response.

## Solution
To mitigate this, this PR rolls back the responses prefect change to
restore stable inline completion behavior.

---

- Treat all work as PUBLIC. Private `feature/x` branches will not be
squash-merged at release time.
- Your code changes must meet the guidelines in
[CONTRIBUTING.md](https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#guidelines).
- License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

jpinkney-aws committed:
test(amazonq): add skipped test report generator (#6540)

## Problem
This scripts creates a report of skipped tests in the amazon q folder

## Solution
Run: `npm run skippedTestReport`

```
Skipped Tests Report
Total skipped tests: 5
Files affected: 4
===================

📁 packages/amazonq/test/e2e/amazonq/featureDev.test.ts
     Skipped tests:
      • /dev {msg} entry (line 165)
      • file-level accepts (line 220)

📁 packages/amazonq/test/e2e/amazonq/template.test.ts
     Skipped tests:
      • Amazon Q Test Template (line 16)

📁 packages/amazonq/test/e2e/amazonq/testGen.test.ts
     Skipped tests:
      • ${language} file (line 168)

📁 packages/amazonq/test/e2e/amazonq/transformByQ.test.ts
     Skipped tests:
      • Running a Java upgrade from start to finish (line 345)
```



---

- Treat all work as PUBLIC. Private `feature/x` branches will not be
squash-merged at release time.
- Your code changes must meet the guidelines in
[CONTRIBUTING.md](https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#guidelines).
- License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

Commit:
Release 3.46.0

Commit:
Release 1.47.0

Commit:
Update version to snapshot version: 1.48.0-SNAPSHOT

Commit:
Update version to snapshot version: 3.47.0-SNAPSHOT

Commit:
Merge release into master

jpinkney-aws committed:
test(core): Use a fake global state for integ/e2e tests (#6565)

## Problem
Features that depend on global state can cause test failures locally
because the tests will use your current global state settings instead of
a fake state for integ/e2e tests

## Solution
integ/e2e tests should use a fake global state instead that
automatically gets reset after each test

---

- Treat all work as PUBLIC. Private `feature/x` branches will not be
squash-merged at release time.
- Your code changes must meet the guidelines in
[CONTRIBUTING.md](https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#guidelines).
- License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

Hweinstock committed:
deps(auth): remove dependence on deprecated and outdated `@aws-sdk/*` packages. (#6474)

## Problem
The auth code relies on old versions of `@aws-sdk/*` that have since
been deprecated or are no longer backward compatible, making versions
bumps impossible.
- `@aws-sdk/credential-provider-imds` has since been
[deprecated](https://www.npmjs.com/package/@aws-sdk/credential-provider-imds)
- `fromIni` from `@aws-sdk/credential-provider-ini` no longer supports
passing a `loadedConfig`.
- `AssumeRoleParams` is no longer exported by
`@aws-sdk/credential-provider-ini`.

We need to be able to bump these `@aws-sdk/*` package versions to
continue to consume newer generated clients. Being pinned to older
versions is also a security risk. See
https://github.com/aws/aws-toolkit-vscode/pull/6439 for more
information.

## Solution
- write custom credentials provider to replace `fromIni` with
`loadedConfig` option.
- drop dependency on `@aws-sdk/credential-provider-ini` since its no
longer used.
- add direct dependency on `@aws-sdk/credential-provider-env` since this
was installed as part of `@aws-sdk-credential-provider-ini` before.
- Fix many (not all) of the deprecation warnings in auth code related to
credentials provider.

### Custom Credentials Provider
Before, we used `fromIni` with the `loadedConfig` option which allows us
to avoid reading the config file from disk on each credentials fetch and
allows us to merge the current credentials with those found in the
`.ini` file. To achieve the same behavior without the `loadedConfig`
option, we need to write our own credentials provider that supports MFA
and role assumption, and returns the desired merged credentials, rather
than reading from disk.

### Testing
- Manually verify this role assumption works by following the steps
[here](https://docs.aws.amazon.com/sdkref/latest/guide/access-assume-role.html).
- Manually verify MFA works via adapting
[this](https://docs.aws.amazon.com/cli/v1/userguide/cli-configure-role.html#:~:text=This%20policy%20allows%20the%20user,they%20authenticate%20by%20using%20MFA.&text=Next%2C%20add%20a%20line%20to,by%20the%20role's%20trust%20policy.&text=The%20mfa_serial%20setting%20can%20take,command%20with%20this%20profile%20fails.&text=The%20second%20profile%20entry%2C%20role,%22:%20%5B%20%7B%20...).
(Used DuoMobile)
- Add unit tests with API calls stubbed. 

## Future Work
- There are two tests that can now be re-enabled because of this version
bump, undoing
https://github.com/aws/aws-toolkit-vscode/commit/db27ebb578c882750e7372b59c37b212f471560c
- The steps to test role assumption could become an integ/e2e test.
Right now requires setting many resources up in console, but perhaps
this can all be done by the SDKs with an account on admin access.

---

- Treat all work as PUBLIC. Private `feature/x` branches will not be
squash-merged at release time.
- Your code changes must meet the guidelines in
[CONTRIBUTING.md](https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#guidelines).
- License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

jpinkney-aws committed:
test(amazonq): Set UI ready before executing amazonq e2e tests (#6580)

## Problem
- after
https://github.com/aws/aws-toolkit-vscode/commit/19d0346db98d45480f9909629b50daf31cba1186
the listener adds messages in a buffer until the ui is ready. It looks
like some messages get caught in the buffer during tests since the mynah
ui activation is slightly different there
- The waitUntil was waiting until an event evaluated to true and then
afterwords checks to see if that same event is still true. This was
causing an issue where if the event originally evaluated to true and
then instantly evaluated to false, your tests would fail

## Solution
- manually set the ui ready after we've created the mynah ui instance in
tests

---

- Treat all work as PUBLIC. Private `feature/x` branches will not be
squash-merged at release time.
- Your code changes must meet the guidelines in
[CONTRIBUTING.md](https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#guidelines).
- License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

Hweinstock committed:
deps(cw): regen clients with codegen at 0.25.x (#6439)

## Problem
The two streaming clients committed to the repo are generated via
different versions of `smithy-typescript-codegen`. Each version of
codegen pins the generated client to use specific versions of
`@aws-sdk/*` and `@smithy/*` packages. The root project also consumes
different versions of these packages. Therefore, the project is
currently consuming three different versions of many `@aws-sdk/*` and
`@smithy/*` packages.

This is problematic because it can cause dependency conflicts. Certain
versions of `@aws-sdk/*` and `@smithy/*` packages are only compatible
with certain other versions. For more information on this, see the
discussion involving help from an SDK team member explaining this
problem.

## Solution
- Regenerate both clients with the same version of codegen `0.25.x`. 
- Pin all necessary versions to ensure a single version of `@aws-sdk/*`
and `@smithy/*` packages.
- Add documentation to the build instructions that these clients must
stay in-sync (in terms of codegen).
- Add documentation regarding updating `@aws-sdk/*` packages. 

---

- Treat all work as PUBLIC. Private `feature/x` branches will not be
squash-merged at release time.
- Your code changes must meet the guidelines in
[CONTRIBUTING.md](https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#guidelines).
- License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

mbfreder committed:
feat(lambda): autoload debug config when Local Invoke view is opened #6583

## Problem
Users have to re-select a debug config every time when open the Invoke
and Debug webview

## Solution
Automatically load the first matching debug configuration when the local
invoke view is opened

leigaol committed:
deps(cw): update SDK clients with codegen at 0.25.x #6584

Update the src.gen dependencies of Amazon Q using new API models.

chungjac committed:
test(amazonq): reset state after each testgen e2e test (#6522)

## Problem
1. The /test command can occasionally create test files in directories
unrelated to the source file location (this is a known science issue
that will be addressed)
2. Unit tests generated after the "Clicks on accept" test may possibly
accumulate in the CI since testFixtures isn't reset for each build. This
can lead to bloated test environments as these tests should be isolated
and start with a clean state.

## Solution
1. Add a dedicated testgen folder within testFixtures to isolate from
other files. This prevents cross-contamination (ex: python tests
appearing in java directories)
2. Clean up generated unit tests after "Clicks on accept" test
---

- Treat all work as PUBLIC. Private `feature/x` branches will not be
squash-merged at release time.
- Your code changes must meet the guidelines in
[CONTRIBUTING.md](https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#guidelines).
- License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

laileni-aws committed:
telemetry(amazonq): Adding requestId for uploadArtifactToS3 in amazonq_utgGenerateTests event (#6598)

## Problem
- IDE does not emit requestId if it fails at `uploadArtifactToS3` in
`/test`

## Solution
- Added requestId for Test Generation without impacting the `/review`
- `requestId` is stored in /test session and emits if something fails in
the test generation workflow.
- Helps in debugging the root cause for the
`UploadTestArtifactToS3Error: "PUT" request failed with code "403"`
error in RTS logs.

---

- Treat all work as PUBLIC. Private `feature/x` branches will not be
squash-merged at release time.
- Your code changes must meet the guidelines in
[CONTRIBUTING.md](https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#guidelines).
- License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

Hweinstock committed:
test(auth): deprecated InputBox usage causes flakiness #6596

## Problem
https://github.com/aws/aws-toolkit-vscode/issues/6594

The `getMfaCodeFromUser` currently uses a deprecated version of
`createInputBox` and the following `promptUser` function:

https://github.com/aws/aws-toolkit-vscode/blob/0fa1b5b47ab642458df934192832a4ffcb6c9d1e/packages/core/src/shared/ui/input.ts#L72-L135

Based on the stack trace, there is a race condition here that can lead
to a type error, but unsure how?
Stack trace: 
```
TypeError: inputBox.hide is not a function
      at promptUser (/Users/runner/work/aws-toolkit-vscode/aws-toolkit-vscode/packages/core/src/shared/ui/input.ts:133:18)
      at async getMfaTokenFromUser (/Users/runner/work/aws-toolkit-vscode/aws-toolkit-vscode/packages/core/src/auth/credentials/utils.ts:124:19)
      at async /Users/runner/work/aws-toolkit-vscode/aws-toolkit-vscode/packages/core/src/auth/providers/sharedCredentialsProvider.ts:422:38
```
### Investigation Notes:
- The error only happens when running the test suite as a whole. 
- The error only happens in CI (unable to reproduce locally). 

## Solution
- avoid use of deprecated `createInputBox`
[function](https://github.com/aws/aws-toolkit-vscode/blob/0fa1b5b47ab642458df934192832a4ffcb6c9d1e/packages/core/src/shared/ui/input.ts#L31-L58),
prefer new version instead.
- Ran the tests 4x with 1000x on the test, and did not see a failure. 

## Future Work 
- Migrate existing cases away from this deprecated `createInputBox` if
there is evidence it leads to flaky tests or unreliable behavior.

avi-alpert committed:
fix(amazonq): Prevent adding same file twice for /doc and /dev (#6568)

## Problem
Stuck on "Uploading code..." for /dev and /doc in a multi-root workspace
with the same file paths in it multiple times.

## Solution
Add a check to prevent same file from being added to archive twice

Fixes issue https://github.com/aws/aws-toolkit-vscode/issues/6566

---

- Treat all work as PUBLIC. Private `feature/x` branches will not be
squash-merged at release time.
- Your code changes must meet the guidelines in
[CONTRIBUTING.md](https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#guidelines).
- License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.

hayemaxi committed:
ci: add post-release notification buildspec (#6609)

Will enable for prod once I confirm it outputs the proper messages.

---

- Treat all work as PUBLIC. Private `feature/x` branches will not be
squash-merged at release time.
- Your code changes must meet the guidelines in
[CONTRIBUTING.md](https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#guidelines).
- License: I confirm that my contribution is made under the terms of the
Apache 2.0 license.
